modules/opengl/glsl/utils/isosurface.glsl

```diff
@@ -0,0 +1,149 @@
+#if MAX_ISOVALUE_COUNT < 1
+#  undef MAX_ISOVALUE_COUNT
+#  define MAX_ISOVALUE_COUNT 1
+#endif
+
+struct IsovalueParameters {
+    float values[MAX_ISOVALUE_COUNT];
+    vec4 colors[MAX_ISOVALUE_COUNT];
+};
+
+/**
+ * Draws an isosurface if the given isovalue is found along the ray in between the 
+ * current and the previous volume sample.
+ *
+ * @param curResult        color accumulated so far during raycasting
+ * @param isovalue         isovalue of isosurface to be drawn
+ * @param isosurfaceColor  color of the isosurface used for blending
+ * @param voxel            scalar values of current sampling position
+ * @param previousVoxel    scalar values of previous sample
+ * @return in case of an isosurface, curResult is blended with the color of the isosurface. 
+ *       Otherwise curResult is returned
+ */
+vec4 drawIsosurface(in vec4 curResult, in float isovalue, in vec4 isosurfaceColor, 
+                    in vec4 voxel, in vec4 previousVoxel,
+                    in sampler3D volume, in VolumeParameters volumeParameters, in int channel, 
+                    in CameraParameters camera, in LightParameters lighting,
+                    in vec3 rayPosition, in vec3 rayDirection, in vec3 toCameraDir,
+                    in float t, in float tIncr, inout float tDepth) {
+    vec4 result = curResult;
+
+    float currentSample = voxel[channel];
+    float prevSample = previousVoxel[channel];
+    float sampleDelta = (currentSample - prevSample);
+
+    // check if the isovalue is lying in between current and previous sample
+    // found isosurface if differences between current/prev sample and isovalue have different signs
+    if ((isovalue - currentSample) * (isovalue - prevSample) <= 0) {
+        // apply linear interpolation between current and previous sample to obtain location of isosurface
+        float a = (currentSample - isovalue) / sampleDelta;
+        // if a == 1, isosurface was already computed for previous sampling position
+        if (a >= 1.0) return result;
+        
+        vec3 isopos = rayPosition - tIncr * a * rayDirection;
+
+        vec4 color = isosurfaceColor;
+#if defined(SHADING_ENABLED)
+        vec3 gradient = COMPUTE_GRADIENT_FOR_CHANNEL(voxel, volume, volumeParameters, isopos, channel);
+        gradient = normalize(gradient);
+
+        // two-sided lighting
+        if (dot(gradient, rayDirection) <= 0) {
+            gradient = -gradient;
+        }
+
+        vec3 isoposWorld = (volumeParameters.textureToWorld * vec4(isopos, 1.0)).xyz;
+        color.rgb = APPLY_LIGHTING(lighting, color.rgb, color.rgb, vec3(1.0),
+                           isoposWorld, -gradient, toCameraDir);
+#endif // SHADING_ENABLED
+
+        // blend surface color and adjust first-hit depth if necessary
+        if (tDepth < 0.0) {
+            // store depth of first hit, i.e. voxel with non-zero alpha
+             tDepth = t;   
+        }
+        color.rgb *= color.a; // use pre-multiplied alpha
```

Comment:
Completely correct would be to composite with the volumetric media before the iso-surface, i.e. taking the distance to the iso-surface, and between, into account.

Reply:
included in latest commit